### PR TITLE
Fixed: `selector-type-no-unknown` is now ignore `/deep/` shadow-piercing descendant combinator.

### PR DESCRIPTION
**Pull request by evilebottnawi**
> Which issue, if any, is this issue related to?

#2506

> Is there anything in the PR that needs further explanation?

No, it's self explanatory.

Maybe need update docs :question: or just add example to `valid` :question:

### CODE REVIEW

File: lib/rules/selector-type-no-unknown/index.js [hunk 1]
```diff
@@ -89,6 +89,7 @@ const rule = function (actual, options) {
             || svgTags.indexOf(tagNameLowerCase) !== -1
             || keywordSets.nonStandardHtmlTags.has(tagNameLowerCase)
             || mathMLTags.indexOf(tagNameLowerCase) !== -1
+            || tagNameLowerCase === "/deep/"
```
Comment (jeddy3) [inline lib/rules/selector-type-no-unknown/index.js]: Let's move this to the [`isStandardSyntaxTypeSelector` util](https://github.com/stylelint/stylelint/blob/master/lib/utils/isStandardSyntaxTypeSelector.js). It's where we build on top of `postcss-selector-parser`.

Also, let's return `false` for all [_reference combinators_](https://www.w3.org/TR/2013/WD-selectors4-20130502/#idref-combinators), not just `/deep/` i.e. return `false` for anything that starts and ends with a `/`.

Be sure to comment the line e.g.
```
// Reference combinators like `/deep/`
```

Comment (evilebottnawi) [inline lib/rules/selector-type-no-unknown/index.js]: @jeddy3 good solution.

Comment (evilebottnawi) [inline lib/rules/selector-type-no-unknown/index.js]: @jeddy3 one remarks: remove tests on invalid `/deeeps/` (i think it is not related to this rules).

File: lib/rules/selector-type-no-unknown/__tests__/index.js [hunk 2]
```diff
@@ -77,6 +77,18 @@ testRule(rule, {
...
+    description: "shadow-piercing descendant combinator",
+  }, {
+    code: ".foo /dEeP/ .bar {}",
+    description: "shadow-piercing descendant combinator",
+  }, {
+    code: ".foo /DEEP/ .bar {}",
+    description: "shadow-piercing descendant combinator",
+  }, {
+    code: ".foo /for/ .bar {}",
+    description: "reference combinators",
```
Comment (davidtheclark) [inline lib/rules/selector-type-no-unknown/__tests__/index.js]: What is a "reference combinator"?

Comment (evilebottnawi) [inline lib/rules/selector-type-no-unknown/__tests__/index.js]: @davidtheclark https://www.w3.org/TR/selectors4/#idref-combinators

Comment (evilebottnawi) [inline lib/rules/selector-type-no-unknown/__tests__/index.js]: @davidtheclark but seem it is deprecated selector, i don't see infomation about this in other spec, btw some project use `/deep/` and be good just ignore this

Comment (davidtheclark) [inline lib/rules/selector-type-no-unknown/__tests__/index.js]: 👍

**Comment by davidtheclark** [comment]
Added to changelog:

> -   Fixed: `selector-type-no-unknown` now ignores the `/deep/` shadow-piercing combinator ([#2508](https://github.com/stylelint/stylelint/pull/2508)).

**Comment by EmilyRosina** [comment]
I still don't understand how to implement this rule in my project :sob: but i need to as we always use `::v-deep` now - HELP :pray:

**Comment by jeddy3** [comment]
@EmilyRosina I don't believe this PR is related to the `::v-deep` pseudo-element.

Please [create a new issue](https://github.com/stylelint/stylelint/issues/new/choose), _following the *bug* template_, if you're getting an unexpected violation.

**Comment by EmilyRosina** [comment]
I just wanted a way to tell the linter that we want to see an error if `/deep/` or `>>>` are used.

**Comment by jeddy3** [comment]
You can use the [`selector-combinator-blacklist` rule](https://github.com/stylelint/stylelint/tree/master/lib/rules/selector-combinator-blacklist) to disallow the `>>>` combinator:

```json
{
  "rules": {
    "selector-combinator-blacklist": [">>>"],
  }
}
```

The `/deep/`[ _reference combinator_](https://www.w3.org/TR/2013/WD-selectors4-20130502/#idref-combinators) is more problematic as the `selector-combinator-blacklist` rule ignores reference combinators. However, you can [write a plugin](https://stylelint.io/developer-guide/plugins) to target and disallow these unusual constructs. I suggest you call the plugin `stylelint-selector-reference-combinator-blacklist`. You can use the code, tests and documentation of the built-in `selector-combinator-blacklist` rule  as a blueprint. Good luck.

**Comment by EmilyRosina** [comment]
Ah ok, so there is no way to do this straight up with stylelint atm :disappointed: 
I guess we'll just rely on Vue/Chrome (unusre who to blame/thank) to tell us for now if someone used `/deep/`. Just a shame we cannot easily do it through the linter.

Thanks anyway :+1:

**Comment by jeddy3** [comment]
> Just a shame we cannot easily do it through the linter.

You can easily do it by writing a plugin. The plugin system exists for exactly these types of scenarios.

**Comment by EmilyRosina** [comment]
It looked a little in-depth on first glance, but maybe it's not as bad as i think :sweat_smile: 
I will have a look once i have more time :smile_cat:

**Pull request merged**